File: supabase/functions/_shared/agent.ts
/**
 * TradvisorAI Agentic Loop — OpenAI Responses API with Grok built-in tools.
 *
 * Ported from Python to TypeScript for Supabase Edge Functions (Deno).
 *
 * Architecture:
 *   User gives high-level intent
 *     -> Grok creates a plan (calls update_plan -> returns to us)
 *     -> Grok searches web + executes code AUTOMATICALLY (server-side)
 *     -> Grok calls update_plan to report progress -> returns to us
 *     -> Repeat until plan is_complete
 *     -> Grok produces final analysis text
 *
 * This version streams Server-Sent Events (SSE) back to the frontend.
 */

import { SYSTEM_PROMPT } from "./prompts.ts";
import { ALL_TOOLS, executeFunction, type PlanData } from "./tools.ts";

// ═══════════════════════════════════════════════════════════════
// SSE EVENT TYPES (sent to frontend)
// ═══════════════════════════════════════════════════════════════

export interface SSEEvent {
  type: "plan_update" | "tool_call" | "step_activity" | "text_delta" | "done" | "error";
  data: Record<string, unknown>;
}

// ═══════════════════════════════════════════════════════════════
// AGENT CONFIGURATION
// ═══════════════════════════════════════════════════════════════

const MAX_ITERATIONS = 20;

function getConfig() {
  const apiKey = Deno.env.get("XAI_API_KEY");
  if (!apiKey) {
    throw new Error("XAI_API_KEY not set in Supabase secrets");
  }
  return {
    apiKey,
    baseUrl: Deno.env.get("XAI_BASE_URL") || "https://api.x.ai/v1",
    model: Deno.env.get("XAI_MODEL") || "grok-3-mini",
  };
}

// ═══════════════════════════════════════════════════════════════
// RESPONSES API CLIENT
// ═══════════════════════════════════════════════════════════════

interface ResponseItem {
  type: string;
  // web_search_call fields
  id?: string;
  status?: string;
  query?: string;
  results?: Array<{ title?: string; url?: string; snippet?: string }>;
  // code_interpreter_call fields
  code?: string;
  language?: string;
  output?: string;
  error?: string;
  // function_call fields
  name?: string;
  arguments?: string;
  call_id?: string;
  // message fields
  content?: Array<{ type: string; text?: string }>;
}

interface APIResponse {
  id: string;
  output: ResponseItem[];
}

async function callResponsesAPI(
  config: ReturnType<typeof getConfig>,
  input: unknown[],
  previousResponseId?: string
): Promise<{ response: APIResponse; usedPreviousId: boolean }> {
  const body: Record<string, unknown> = {
    model: config.model,
    tools: ALL_TOOLS,
    input,
  };

  if (previousResponseId) {
    body.previous_response_id = previousResponseId;
  }

  const res = await fetch(`${config.baseUrl}/responses`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${config.apiKey}`,
    },
    body: JSON.stringify(body),
  });

  if (!res.ok) {
    const errorText = await res.text();

    // If error is about image media_type (accumulated images in context),
    // retry without previous_response_id to start fresh context
    if (previousResponseId && errorText.includes("media_type")) {
      console.warn("Retrying without previous_response_id due to image media_type error");
      const retryBody: Record<string, unknown> = {
        model: config.model,
        tools: ALL_TOOLS,
        input,
      };
      const retryRes = await fetch(`${config.baseUrl}/responses`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${config.apiKey}`,
        },
        body: JSON.stringify(retryBody),
      });

      if (!retryRes.ok) {
        const retryErrorText = await retryRes.text();
        throw new Error(`xAI API error (${retryRes.status}): ${retryErrorText}`);
      }

      return { response: (await retryRes.json()) as APIResponse, usedPreviousId: false };
    }

    throw new Error(`xAI API error (${res.status}): ${errorText}`);
  }

  return { response: (await res.json()) as APIResponse, usedPreviousId: !!previousResponseId };
}

// ═══════════════════════════════════════════════════════════════
// AGENT RUNNER (GENERATOR)
// ═══════════════════════════════════════════════════════════════

/**
 * Run the full agentic loop for a user query.
 * Yields SSE events as the agent works.
 *
 * @param userQuery - The user's message
 * @param history - Previous messages in the conversation (for context)
 */
export async function* runAgent(
  userQuery: string,
  history: Array<{ role: string; content: string }> = []
): AsyncGenerator<SSEEvent> {
  const config = getConfig();
  let plan: PlanData | null = null;
  let responseId: string | undefined;
  let fullResponseText = "";

  // Build input messages — ensure all content is plain text strings
  // to avoid sending malformed image blocks without media_type
  const sanitizedHistory = history
    .filter((msg) => msg.content && typeof msg.content === "string")
    .map((msg) => ({
      role: msg.role,
      content: typeof msg.content === "string" ? msg.content : String(msg.content),
    }));

  const inputMessages: unknown[] = [
    { role: "system", content: SYSTEM_PROMPT },
    ...sanitizedHistory,
    { role: "user", content: userQuery },
  ];

  for (let iteration = 1; iteration <= MAX_ITERATIONS; iteration++) {
    // ── Call the Responses API ──────────────────────
    let response: APIResponse;
    try {
      const result = await callResponsesAPI(config, inputMessages, responseId);
      response = result.response;
      responseId = response.id;
    } catch (err) {
      yield {
        type: "error",
        data: { message: `API Error: ${(err as Error).message}` },
      };
      yield {
        type: "done",
        data: { iterations: iteration, plan },
      };
      return;
    }

    // ── Parse output items ──────────────────────────
    const functionCallOutputs: unknown[] = [];

    for (const item of response.output) {
      // --- Built-in: web search executed server-side ---
      if (item.type === "web_search_call") {
        yield {
          type: "tool_call",
          data: { name: "web_search", description: item.query || "Searching the web..." },
        };

        // Emit search query as activity
        if (item.query) {
          yield {
            type: "step_activity",
            data: {
              activity_type: "search",
              content: item.query,
              metadata: { search_query: item.query }
            }
          };
        }

        // Emit search results as activities
        if (item.results && item.results.length > 0) {
          const summary = item.results.slice(0, 3).map(r => r.title || r.snippet).filter(Boolean).join(", ");
          if (summary) {
            yield {
              type: "step_activity",
              data: {
                activity_type: "info",
                content: `Found: ${summary.substring(0, 150)}${summary.length > 150 ? "..." : ""}`
              }
            };
          }
        }
      }

      // --- Built-in: code interpreter executed server-side ---
      else if (item.type === "code_interpreter_call") {
        yield {
          type: "tool_call",
          data: {
            name: "code_execution",
            description: item.code
              ? item.code.split("\n").find((l: string) => l.trim())?.substring(0, 70) || "Running analysis..."
              : "Running analysis...",
          },
        };

        // Emit code as activity
        if (item.code) {
          yield {
            type: "step_activity",
            data: {
              activity_type: "code",
              content: item.code,
              metadata: { language: item.language || "python" }
            }
          };
        }

        // Emit output as activity
        if (item.output) {
          yield {
            type: "step_activity",
            data: {
              activity_type: "output",
              content: item.output.substring(0, 300) // Limit output length
            }
          };
        }

        // Emit error as activity
        if (item.error) {
          yield {
            type: "step_activity",
            data: {
              activity_type: "info",
              content: `⚠️ Error: ${item.error}`
            }
          };
        }
      }

      // --- Custom function call (update_plan) ---
      else if (item.type === "function_call") {
        const name = item.name!;
        const args = item.arguments || "{}";
        const callId = item.call_id!;

        let argsDict: Record<string, unknown> = {};
        try {
          argsDict = JSON.parse(args);
        } catch {
          // ignore parse errors
        }

        // Emit tool call event
        yield {
          type: "tool_call",
          data: {
            name,
            description: (argsDict.task_summary as string) || "",
          },
        };

        // Track plan state
        if (name === "update_plan") {
          plan = argsDict as unknown as PlanData;
          yield {
            type: "plan_update",
            data: {
              task_summary: argsDict.task_summary || "",
              steps: argsDict.steps || [],
              is_complete: argsDict.is_complete || false,
            },
          };
        }

        // Execute the custom function
        const result = executeFunction(name, args);

        // Queue the result to send back
        functionCallOutputs.push({
          type: "function_call_output",
          call_id: callId,
          output: result,
        });
      }

      // --- Text message from the model ---
      else if (item.type === "message") {
        for (const part of item.content || []) {
          if (part.type === "output_text" && part.text) {
            fullResponseText += part.text;
            yield {
              type: "text_delta",
              data: { content: part.text },
            };
          }
        }
      }
    }

    // ── Decide whether to continue ──────────────────
    if (functionCallOutputs.length === 0) {
      // No custom function calls -> model is done
      yield {
        type: "done",
        data: { iterations: iteration, plan, fullText: fullResponseText },
      };
      return;
    }

    // Send function results back and continue the loop
    inputMessages.length = 0;
    inputMessages.push(...functionCallOutputs);
  }

  // Max iterations reached
  yield {
    type: "text_delta",
    data: { content: "\n\n(Reached maximum iterations.)" },
  };
  yield {
    type: "done",
    data: { iterations: MAX_ITERATIONS, plan, fullText: fullResponseText },
  };
}

File: agent/agent.py
"""
TradvisorAI Agentic Loop — OpenAI Responses API with Grok built-in tools.

Key upgrade from v1:
  - web_search runs SERVER-SIDE on Grok (browses actual pages, not just snippets)
  - code_interpreter runs SERVER-SIDE (real sandbox with numpy/pandas)
  - Stateful conversations via previous_response_id (cheaper, automatic caching)
  - Only custom functions (update_plan) are handled locally

Architecture:
  User gives high-level intent
    → Grok creates a plan (calls update_plan → returns to us)
    → Grok searches web + executes code AUTOMATICALLY (server-side)
    → Grok calls update_plan to report progress → returns to us
    → Repeat until plan is_complete
    → Grok produces final analysis text
"""

import json
from dataclasses import dataclass, field
from typing import Generator

from openai import OpenAI

from config import XAI_API_KEY, BASE_URL, MODEL, MAX_ITERATIONS
from prompts import SYSTEM_PROMPT
from tools import ALL_TOOLS, execute_function


# ═══════════════════════════════════════════════════════════════
# EVENT TYPES (yielded by the agent to the UI layer)
# ═══════════════════════════════════════════════════════════════


@dataclass
class PlanUpdate:
    """Agent created or updated its execution plan."""
    task_summary: str
    steps: list
    is_complete: bool
    explanation: str = ""


@dataclass
class ToolCall:
    """Agent is calling a tool (built-in or custom)."""
    name: str
    description: str = ""


@dataclass
class TextDelta:
    """Text chunk from the agent's response."""
    content: str


@dataclass
class Done:
    """Agent finished."""
    iterations: int
    plan: dict | None


# ═══════════════════════════════════════════════════════════════
# AGENT
# ═══════════════════════════════════════════════════════════════


class TradvisorAgent:
    """
    Agentic loop using the Responses API.

    Built-in tools (web_search, code_interpreter) execute automatically on
    Grok's servers.  Only custom function calls (update_plan) return to us
    for handling, creating the Cursor-style plan-update loop.
    """

    def __init__(self):
        self.client = OpenAI(
            api_key=XAI_API_KEY,
            base_url=BASE_URL,
        )
        self.plan: dict | None = None
        self.response_id: str | None = None

    # ── public ──────────────────────────────────────────────

    def run(self, user_query: str) -> Generator:
        """
        Run the full agentic loop for a user query.
        Yields events as the agent works.
        """
        self.plan = None
        self.response_id = None

        # First call — full input with system + user message
        input_messages = [
            {"role": "system", "content": SYSTEM_PROMPT},
            {"role": "user", "content": user_query},
        ]

        for iteration in range(1, MAX_ITERATIONS + 1):
            # ── Call the Responses API ──────────────────────
            try:
                kwargs = {
                    "model": MODEL,
                    "tools": ALL_TOOLS,
                    "input": input_messages,
                }
                # Continue stateful conversation if we have a previous response
                if self.response_id:
                    kwargs["previous_response_id"] = self.response_id

                response = self.client.responses.create(**kwargs)
                self.response_id = response.id

            except Exception as e:
                yield TextDelta(f"\n\nAPI Error: {e}")
                yield Done(iterations=iteration, plan=self.plan)
                return

            # ── Parse output items ──────────────────────────
            function_call_outputs = []

            for item in response.output:
                item_type = item.type

                # --- Built-in: web search executed server-side ---
                if item_type == "web_search_call":
                    yield ToolCall(name="web_search", description="Searching the web...")

                # --- Built-in: code interpreter executed server-side ---
                elif item_type == "code_interpreter_call":
                    yield ToolCall(name="code_execution", description="Executing Python code...")

                # --- Custom function call (update_plan) ---
                elif item_type == "function_call":
                    name = item.name
                    arguments = item.arguments
                    call_id = item.call_id

                    try:
                        args_dict = json.loads(arguments)
                    except json.JSONDecodeError:
                        args_dict = {}

                    yield ToolCall(name=name, description=args_dict.get("task_summary", ""))

                    # Track plan state
                    if name == "update_plan":
                        self.plan = args_dict
                        yield PlanUpdate(
                            task_summary=args_dict.get("task_summary", ""),
                            steps=args_dict.get("steps", []),
                            is_complete=args_dict.get("is_complete", False),
                            explanation=args_dict.get("explanation", "") or "",
                        )

                    # Execute the custom function
                    result = execute_function(name, arguments)

                    # Queue the result to send back
                    function_call_outputs.append({
                        "type": "function_call_output",
                        "call_id": call_id,
                        "output": result,
                    })

                # --- Text message from the model ---
                elif item_type == "message":
                    for content_part in getattr(item, "content", []):
                        if getattr(content_part, "type", "") == "output_text":
                            yield TextDelta(content_part.text)

            # ── Decide whether to continue ──────────────────
            if not function_call_outputs:
                # No custom function calls → model is done
                yield Done(iterations=iteration, plan=self.plan)
                return

            # Send function results back and continue the loop
            input_messages = function_call_outputs

        # Max iterations
        yield TextDelta("\n\n(Reached maximum iterations.)")
        yield Done(iterations=MAX_ITERATIONS, plan=self.plan)


def run_agent(query: str) -> Generator:
    """Convenience function to run the agent."""
    agent = TradvisorAgent()
    return agent.run(query)
